src/driver/mongodb.rs

```diff
@@ -0,0 +1,248 @@
+      ))
+    }
+  }
+  
+  fn validate_definition(definition: &Definition) -> Result<(), Error> {
+    match &definition.data {
+      &Schema::Object{ref properties,..} => {
+        for (key, value) in properties {
+          match value {
+            &Schema::Array{ref items} => {
+              match items.deref() {
+                &Schema::Object{..} => (),
+                _ => return Err(Error::validation(
+                  format!("Items for collection '{}' must be an object.", key),
+                  format!("The MongoDB driver only supports inserting objects into collections, try changing the schema type of the '{}' collection to reflect this.", key)
+                ))
+              }
+            },
+            _ => return Err(Error::validation(
+              format!("Schema type for the '{}' property must be an array.", key),
+              format!("The MongoDB driver only supports array collections as the first level data types, try changing the '{}' property in your schema to reflect this.", key)
+            ))
+          }
+        }
+      },
+      _ => return Err(Error::validation(
+        "Root data schema is not object.",
+        "The MongoDB driver only supports an object as the root data type, try changing your schema to reflect this."
+      ))
+    }
+    Ok(())
+  }
+  
+  // TODO: Better inline documentation.
+  fn query(&self, query: Query) -> Result<Value, Error> {
+    match query {
+      Query::Value => Err(Error::new(
+        ErrorCode::Forbidden,
+        "Can’t query the entire MongoDB database.",
+        Some("Query something more specfic instead of the entire database.")
+      )),
+      Query::Object(collection_queries) => {
+        // First level is the collection.
+        let mut object = LinearMap::new();
+        for (Selection::Key(coll_name), query) in collection_queries {
+          // if let Selection::Key(coll_name) = selection {
+            let collection = self.db.collection(&coll_name);
+            match query {
+              // TODO: Make this a range error when implementing selection by
+              // range.
+              Query::Value => {
+                let mut cursor = try!(collection.find(None, None));
+                let mut values = Vec::new();
+                if let Some(Err(error)) = cursor.find(|entry| match entry {
+                  &Ok(ref document) => { values.push(Value::from(document.clone())); false },
+                  &Err(_) => true
+                }) {
+                  return Err(Error::from(error));
+                } else {
+                  object.insert(coll_name, Value::Array(values));
+                }
+              },
+              // TODO: When implementing collections consider not using the
+              // MongoDB `_id` property as the key.
+              Query::Object(_) => {
+                
+              }
+            }
+          // } else {
```

Comment:
It's code which should really be there, but because there is only one variant in `Selection` an error is thrown when it is there. This will change shortly.